### ISSUE TITLE
[FEA] Get Series.list offsets / Construct Series of lists from offsets and values

### ISSUE THREAD

**Issue by bdice** [rapidsai/cudf#11130]
**Is your feature request related to a problem? Please describe.**
I would like to be able to access the offsets of a Series of lists. That would allow me to implement a function like `list_add` that takes two "awkward arrays," Series of lists of numbers that have the same list shape, and adds them together. The binary operation can be straightforwardly applied to the "leaves" of each list column, which is the child column containing the data. However, to do this, I need a way to access indices and rebuild the list structure. For example, if `Series.list.offsets` and `cudf.Series.list.from_arrays(offsets, values)` existed, I could run something like:

```python
def list_add(s1, s2):
    """Take two Series of lists of numerical data and add them."""
    # Ignore nested lists for simplicity -- this only works for a single level of lists
    if s1.list.offsets!= s2.list.offsets:
        raise ValueError("List columns must have corresponding offsets.")
    return cudf.Series.list.from_arrays(s1.list.offsets, s1.list.leaves + s2.list.leaves)
```

**Describe the solution you'd like**
- Implement a property `Series.list.offsets` that exposes the offset array, similar to [PyArrow's `pyarrow.ListArray.offsets`](https://arrow.apache.org/docs/python/generated/pyarrow.ListArray.html#pyarrow.ListArray.offsets) but returning a GPU-resident array.

- Implement a constructor `Series.list.from_arrays(offsets, values)` that builds a Series of lists from input offsets and values, similar to [PyArrow's `pyarrow.ListArray.from_arrays`](https://arrow.apache.org/docs/python/generated/pyarrow.ListArray.html#pyarrow.ListArray.from_arrays) but enabling construction from GPU-resident arrays.

**Describe alternatives you've considered**
I strongly prefer this approach over implementing binops directly on list types because it allows for precise control of what APIs are exposed and how they behave. Implementing binops for lists would allow for operators like `+` to be used, which is prone to error because it overloads the Python-like list semantics of "adding is list concatenation" with the array-like semantics of normal addition.

**Additional context**
It's not clear to me where the name "leaves" came from. To align with PyArrow, we would rename "leaves" to `Series.list.values`.

**Comment by bdice** [comment]
@shwina That's very helpful, I did not consider `explode()`/`agg(list)`. To simplify and match this example:
```python
def list_add(s1, s2):
    return (s1.explode() - s2.explode()).groupby(level=0).agg(list)
```
I'm guessing that `explode()` returns a copy, unlike `Series.list.leaves`, and that `groupby(level=0).agg(list)` is nontrivial to compute compared to a constructor from offsets and values. Perhaps there would be good reasons for performance and flexibility to expose the offset accessor / list constructor primitives.

As to whether offsets are an implementation detail -- I considered this as well. My view is that offsets are helpful to expose and doing so does not make stronger promises about our data model than what we already make in other ways ([offsets are already exposed in the libcudf API](https://docs.rapids.ai/api/libcudf/nightly/classcudf_1_1lists__column__view.html#a3d738360310aad73666f1e3de3b7e50e), and cuDF has a stated aim to be Arrow-conformant to a large extent).

I definitely agree that the ability to do this computation in a zero-copy way and compatibility with GPU Awkward Arrays would be desirable. Exposing the raw offsets and a way to rebuild a list from them seems like a good step in both of those directions.

**Comment by shwina** [comment]
> It's not clear to me where the name "leaves" came from. To align with PyArrow, we would rename "leaves" to Series.list.values.

Note  that  `values` are distinct from `leaves`:

The `values` of a list array is what you get by removing "one level of nesting" from the array:

```python
>>> pa.array([[[[1, 2]]]]).values
<pyarrow.lib.ListArray object at 0x7fe4de05df40>
[
  [
    [
      1,
      2
    ]
  ]
]
```

Whereas what we call `leaves` is what you get from removing _all_ levels of nesting:

```python
In [7]: cudf.Series([[[[1, 2]]]]).list.leaves
Out[7]:
0    1
1    2
dtype: int64
```

**Comment by bdice** [comment]
@shwina Interesting. Would you consider exposing both `list.values` and `list.leaves`? It seems important to have a way to un-nest one level at a time (like with `list.offsets`).

**Comment by shwina** [comment]
Right, which is why I'm suggesting a distinct `DataFrame.list.eval` API (note the namespace).

**Comment by bdice** [comment]
> Right, which is why I'm suggesting a distinct `DataFrame.list.eval` API (note the namespace).

I missed that namespace, thanks for the pointer. I have a lot of questions about how this would act and I don’t think the answers are obvious. AST limitations could be harshly constraining here and no broadcasting would be possible. It also introduces an undesirable asymmetry between operators and eval, and is beyond the API scope of both Pandas and Arrow…

**Comment by GregoryKimball** [comment]
Hello @shwina and @bdice, [bucketize](https://github.com/pytorch/torcharrow/blob/main/csrc/velox/functions/rec/bucketize.h) is a feature that we might unlock if we could construct a list column from offsets and values. 
Bucketize is performed on leaves and uses the same offsets of the input columns.

```
def bucketize(series, buckets):    
    ans = cudf.Series([0] * len(series.list.leaves)) 
    for b in buckets:
        ans += series.list.leaves > b
    return cudf.Series.list.from_arrays(series.list.offsets, ans.list.leaves)
```

**Comment by GregoryKimball** [comment]
To my surprise the `explode` trick from #10967 works here as well:
```
def bucketize(a, buckets):
    # make an exploded table with the offsets from 'a' and
    # zeros for the values
    b = (a.explode() - a.explode()).groupby(level=0).agg(list)    
    # add one for each bucket boundary
    for k in buckets:
        b = (b.explode() + (a.explode() > k)).groupby(level=0).agg(list)
    return b
```

```
                 a                b
0     [1, 2, 3, 3]     [0, 1, 2, 2]
1  [1, 2, 1, 0, 1]  [0, 1, 0, 0, 0]
```

**Comment by shwina** [comment]
Then `list.bucketize()` is an API we may want to consider adding, rather than having each user write their own version of it.

**Comment by VibhuJawa** [comment]
Documenting another usecase that `Series.list.from_arrays` will solve for us is going easily b/w a list of token id arrays and saving it into cuDF list series. This is very common for NLP workflows. 

The input of that is often a batch of token ids (either as input or output).  

Currently the workflow looks like below: 

Input:
```python3
# Example usage
token_batch1 = cp.array([[1, 2, 1], [1, 1, 0]])
token_batch2 = cp.array([[5, 6]])
token_batch3 = cp.array([[7, 8, 9]])
```

Gather offsets: 
```python3
def flatten_arrays_and_offsets(array_list):
    flattened_array = cp.concatenate([arr.flatten() for arr in array_list])    
    row_sizes = [arr.shape[1] for arr in array_list for _ in range(arr.shape[0])]
    cumulative_offsets = cp.cumsum(cp.concatenate([cp.array([0]), cp.array(row_sizes)]))    
    return flattened_array, cumulative_offsets
    
flattened_array, offsets = flatten_arrays_and_offsets([token_batch1, token_batch2, token_batch3])
#flattened_array:  array([1, 2, 1, 1, 1, 0, 5, 6, 7, 8, 9])
# offsets: array([ 0,  3,  6,  8, 11])

```

Create a list column from offsets:
```python3
def create_2d_list_from_ar_and_offsets(offsets, flattened_array):
    data = as_column(flattened_array)
    offset_col = as_column(offsets.astype(cp.int32))
    n_rows = len(offsets)-1
    mask_col = cp.full(shape=n_rows, fill_value=cp.bool_(True))
    mask = cudf._lib.transform.bools_to_mask(as_column(mask_col))
    lc = cudf.core.column.ListColumn(
        size=n_rows,
        dtype=cudf.ListDtype(data.dtype),
        mask=mask,
        offset=0,
        null_count=0,
        children=(offset_col, data),
    )
    
    return cudf.Series(lc)

``` 

```python3
create_2d_list_from_ar_and_offsets(offsets, flattened_array)

0    [1, 2, 1]
1    [1, 1, 0]
2       [5, 6]
3    [7, 8, 9]
dtype: list
```

What i think will help here is this constructor that @bdice  mentioned. 

```python3
Series.list.from_arrays(offsets, values)
```

Because now the code will just look like below (instead of user having to reach into `cuDF`) and create the column using `create_2d_list_from_ar_and_offsets` function. 

```python3
s = Series.list.from_arrays(offsets=offsets, values=flattened_array)
```

**Comment by bdice** [comment]
Related project for handling nested data (supports cudf): https://github.com/intake/akimbo
